File: frontend/src/api/videoAssembly.ts
/**
 * Phase 10 — Video Assembly Engine API client.
 * Mirrors musicEngine.ts structure exactly.
 */
import apiClient from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface VideoAssemblyJob {
  id: string;
  project_id: string;
  episode_id: string | null;
  job_type: string;
  status: "pending" | "running" | "completed" | "failed";
  mode: string;
  triggered_by: string;
  params: Record<string, unknown>;
  result: Record<string, unknown> | null;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  duration_seconds: number | null;
  created_at: string;
  updated_at: string;
}

export interface VideoOutput {
  id: string;
  job_id: string;
  project_id: string;
  episode_id: string | null;
  output_type: string;
  status: string;
  storage_key: string;
  storage_bucket: string;
  file_size_bytes: number;
  duration_seconds: number;
  width: number;
  height: number;
  fps: number;
  format: string;
  provider: string;
  scene_count: number;
  has_voice: boolean;
  has_music: boolean;
  has_subtitles: boolean;
  quality_passed: boolean;
  quality_score: number;
  output_metadata: Record<string, unknown>;
  created_at: string;
}

export interface VideoRetryEntry {
  id: string;
  project_id: string;
  episode_id: string | null;
  original_job_id: string | null;
  status: string;
  retry_count: number;
  max_retries: number;
  reason: string;
  params: Record<string, unknown>;
  next_retry_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface PaginationMeta {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface VideoAssemblyDashboardStats {
  total_jobs: number;
  jobs_completed: number;
  jobs_pending: number;
  jobs_failed: number;
  jobs_running: number;
  total_video_outputs: number;
  total_retry_entries: number;
  recent_jobs: VideoAssemblyJob[];
}

export interface DispatchResponse {
  job_id: string;
  task_id: string;
  mode: string;
  status: string;
}

export interface TriggerAssembleEpisodeRequest {
  project_id: string;
  episode_id?: string | null;
  output_type?: "episode_cut" | "short_form_cut";
  width?: number;
  height?: number;
  fps?: number;
  include_subtitles?: boolean;
  triggered_by?: string;
}

// ---------------------------------------------------------------------------
// API
// ---------------------------------------------------------------------------

const BASE = '/va'

export const videoAssemblyApi = {
  getDashboard: (projectId: string) =>
    apiClient
      .get<VideoAssemblyDashboardStats>(`${BASE}/dashboard/${projectId}`)
      .then((r) => r.data),

  listJobs: (
    projectId: string,
    params?: { page?: number; page_size?: number; status?: string }
  ) =>
    apiClient
      .get<{ items: VideoAssemblyJob[]; meta: PaginationMeta }>(`${BASE}/jobs`, {
        params: { project_id: projectId, ...params },
      })
      .then((r) => r.data),

  getJob: (jobId: string) =>
    apiClient.get<VideoAssemblyJob>(`${BASE}/jobs/${jobId}`).then((r) => r.data),

  listOutputs: (
    projectId: string,
    params?: { page?: number; page_size?: number; output_type?: string }
  ) =>
    apiClient
      .get<{ items: VideoOutput[]; meta: PaginationMeta }>(`${BASE}/outputs`, {
        params: { project_id: projectId, ...params },
      })
      .then((r) => r.data),

  getOutput: (outputId: string) =>
    apiClient.get<VideoOutput>(`${BASE}/outputs/${outputId}`).then((r) => r.data),

  listRetryQueue: (projectId: string, params?: { page?: number; page_size?: number }) =>
    apiClient
      .get<{ items: VideoRetryEntry[]; meta: PaginationMeta }>(`${BASE}/retry-queue`, {
        params: { project_id: projectId, ...params },
      })
      .then((r) => r.data),

  triggerAssembleEpisode: (req: TriggerAssembleEpisodeRequest) =>
    apiClient
      .post<DispatchResponse>(`${BASE}/generate/assemble-episode`, req)
      .then((r) => r.data),

  sweepRetryQueue: (projectId: string, limit = 10) =>
    apiClient
      .post<DispatchResponse>(`${BASE}/retry-queue/sweep`, { project_id: projectId, limit })
      .then((r) => r.data),
};

File: backend/apps/api/schemas/video_assembly.py
"""
Phase 10 — Video Assembly Engine Pydantic schemas.

Mirrors apps/api/schemas/music_engine.py exactly.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Shared
# ---------------------------------------------------------------------------

class PaginationMeta(BaseModel):
    total: int
    page: int
    page_size: int
    total_pages: int


# ---------------------------------------------------------------------------
# Job
# ---------------------------------------------------------------------------

class VideoAssemblyJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    episode_id: uuid.UUID | None
    job_type: str
    status: str
    mode: str
    triggered_by: str
    params: dict[str, Any]
    result: dict[str, Any] | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    duration_seconds: float | None
    created_at: datetime
    updated_at: datetime


class VideoAssemblyJobListResponse(BaseModel):
    items: list[VideoAssemblyJobResponse]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Output
# ---------------------------------------------------------------------------

class VideoOutputResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    job_id: uuid.UUID
    project_id: uuid.UUID
    episode_id: uuid.UUID | None
    output_type: str
    status: str
    storage_key: str
    storage_bucket: str
    file_size_bytes: int
    duration_seconds: float
    width: int
    height: int
    fps: int
    format: str
    provider: str
    scene_count: int
    has_voice: bool
    has_music: bool
    has_subtitles: bool
    quality_passed: bool
    quality_score: float
    output_metadata: dict[str, Any]
    created_at: datetime


class VideoOutputListResponse(BaseModel):
    items: list[VideoOutputResponse]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Retry queue
# ---------------------------------------------------------------------------

class VideoRetryQueueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    episode_id: uuid.UUID | None
    original_job_id: uuid.UUID | None
    status: str
    retry_count: int
    max_retries: int
    reason: str
    params: dict[str, Any]
    next_retry_at: datetime | None
    created_at: datetime
    updated_at: datetime


class VideoRetryQueueListResponse(BaseModel):
    items: list[VideoRetryQueueResponse]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Dashboard stats
# ---------------------------------------------------------------------------

class VideoAssemblyDashboardStats(BaseModel):
    total_jobs: int
    jobs_completed: int
    jobs_pending: int
    jobs_failed: int
    jobs_running: int
    total_video_outputs: int
    total_retry_entries: int
    recent_jobs: list[VideoAssemblyJobResponse]


# ---------------------------------------------------------------------------
# Trigger requests
# ---------------------------------------------------------------------------

class TriggerAssembleEpisodeRequest(BaseModel):
    project_id: uuid.UUID
    episode_id: uuid.UUID | None = None
    output_type: str = "episode_cut"   # "episode_cut" | "short_form_cut"
    width: int = 1920
    height: int = 1080
    fps: int = 24
    include_subtitles: bool = False
    triggered_by: str = "api"


class TriggerRetryQueueRequest(BaseModel):
    project_id: uuid.UUID
    limit: int = 10


# ---------------------------------------------------------------------------
# Dispatch response
# ---------------------------------------------------------------------------

class DispatchResponse(BaseModel):
    job_id: str
    task_id: str
    mode: str
    status: str
